acme/src/acme/client.py

```diff
@@ -280,6 +282,15 @@ def poll_finalization(self, orderr: messages.OrderResource,
                     alt_chains = [self._post_as_get(url).text for url in alt_chains_urls]
                     orderr = orderr.update(alternative_fullchains_pem=alt_chains)
                 return orderr
+            else:
+                assert body.status == messages.STATUS_PROCESSING
+                # the only other option is messages.STATUS_PENDING, which we should never hit here
+                # because we've already checked for it earlier
```

Comment:
how do we know the status cannot be pending here?

there's a good chance i'm just missing something, but the only other reference to pending i see is [here](https://github.com/certbot/certbot/blob/723fe64d4dbc778f1e8fd0b93320c1a30dc6fa95/acme/src/acme/client.py#L202) which i believe is about authorizations. can we be confident that orders atomically move out of the pending state when all authorizations are valid? even so, if this is the check you're referring to, i (maybe unnecessarily) worry that things could break in the future as it relies on behavior across functions and the order they were called

whatcha think?

Reply:
No, you're right, this could happen non-atomically. I was conflating _authorization_ pending checks with _order_ pending checks. And on that, the spec says:

> Order objects are created in the "pending" state.  Once all of the
>  authorizations listed in the order object are in the "valid" state,
>  the order transitions to the "ready" state.  

Which should be quick, and we do wait a second, but technically isn't required to be atomic. I'll get rid of the assert; the code should otherwise work as-is. Or I could check the status before checking for a retry-after, which I think should only be there when it's in processing, but the code should behave the same either way.